cmd/kube-controller-manager/app/core.go

```diff
@@ -603,9 +603,26 @@ func setNodeCIDRMaskSizes(cfg nodeipamconfig.NodeIPAMControllerConfiguration) (i
 // setNodeCIDRMaskSizesDualStack returns the IPv4 and IPv6 node cidr mask sizes to the value provided
 // for --node-cidr-mask-size-ipv4 and --node-cidr-mask-size-ipv6 respectively. If value not provided,
 // then it will return default IPv4 and IPv6 cidr mask sizes.
-func setNodeCIDRMaskSizesDualStack(cfg nodeipamconfig.NodeIPAMControllerConfiguration) (int, int, error) {
+func setNodeCIDRMaskSizesDualStack(cfg nodeipamconfig.NodeIPAMControllerConfiguration, serviceCIDR, secondaryServiceCIDR *net.IPNet) (int, int, error) {
 	ipv4Mask, ipv6Mask := defaultNodeMaskCIDRIPv4, defaultNodeMaskCIDRIPv6
-	// NodeCIDRMaskSize can be used only for single stack clusters
+	// compatible with existing single-stack clusters when dual-stack feature gate become enabled by default
+	if cfg.NodeCIDRMaskSize != 0 && cfg.NodeCIDRMaskSizeIPv4 == 0 && cfg.NodeCIDRMaskSizeIPv6 == 0 {
+		if secondaryServiceCIDR != nil {
+			return ipv4Mask, ipv6Mask, errors.New("usage of --node-cidr-mask-size is not allowed with dual-stack clusters")
+		}
+		if serviceCIDR != nil {
+			if serviceCIDR.IP.To4() != nil {
+				ipv4Mask = int(cfg.NodeCIDRMaskSize)
+				klog.Warningf("--node-cidr-mask-size is specified when dual-stack feature gate enabled, using %v as the mask size for IPv4 node cidr", ipv4Mask)
```

Comment:
This should not be a warning. It should be 100% allowed to use the old argument name in any cluster with a single-stack configuration, regardless of the feature gate setting.